assets/model_monitoring/components/src/action_analyzer/contracts/action_detectors/metrics_violation_index_action_detector.py

```diff
@@ -48,16 +67,86 @@ def preprocess_data(self, df: pandas.DataFrame) -> pandas.DataFrame:
         Returns:
             pandas.DataFrame: preprocessed pandas dataframe.
         """
-        pass
+        try:
+            print("Start to run MetricsViolationIndexActionDetector.")
+            preprocessed_df = extract_fields_from_debugging_info(df, self.index_id)
+            return preprocessed_df
+        except Exception as e:
+            print("MetricsViolationIndexActionDetector preprocess failed with error", e)
+            return pandas.DataFrame()
 
-    def detect(self, df: pandas.DataFrame, llm_client: LLMClient) -> list(Action):
+    def detect(self, df: pandas.DataFrame, llm_client: LLMClient, aml_deployment_id=None) -> List[Action]:
         """Detect the action.
 
         Args:
             df(pandas.DataFrame): input pandas dataframe.
             llm_client(LLMClient): LLM client used to get some llm scores/info for action.
+            aml_deployment_id(str): (Optional) aml deployment id for the action.
+
+        Returns:
+            List[Action]: list of actions.
+        """
+        action_list = []
+        try:
+            # get llm retrieval score
+            df[INDEX_SCORE_LLM_COLUMN] = df.apply(get_retrieval_score, axis=1, args=(llm_client,))
```

Comment:
call twice may cause double cost to the customer

Reply:
refactored